source/agora/test/PreimageSharing.d

```diff
@@ -0,0 +1,63 @@
+/*******************************************************************************
+
+    Contains tests for sharing preimages during SCP protocol run.
+
+    Copyright:
+        Copyright (c) 2019-2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.PreimageSharing;
+
+version (unittest):
+
+import agora.common.Config;
+import agora.common.ManagedDatabase;
+import agora.consensus.data.Block;
+import agora.consensus.data.Params;
+import agora.consensus.data.PreImageInfo;
+import agora.consensus.data.Transaction;
+import agora.consensus.EnrollmentManager;
+import agora.consensus.state.UTXOSet;
+import agora.crypto.Schnorr;
+import agora.crypto.Key;
+import agora.utils.Test;
+import agora.test.Base;
+
+import std.exception;
+import std.path : buildPath;
+
+import core.atomic;
+import core.stdc.stdint;
+import core.thread;
+
+import geod24.Registry;
+
+unittest
+{
+    TestConf conf = {
+        quorum_threshold : 100,
+        preimage_catchup_interval : 1.seconds,
+    };
+
+    // set up nodes
+    auto network = makeTestNetwork!LazyAPIManager(conf);
```

Comment:
You can see in the comment why it works
```D
/// Please note that preimages can still be shared during nomination, if the
/// TestConf.preimage_catchup_interval is set to a low value.
public class NoPreImageVN : TestValidatorNode
```

Reply:
But then, the other test that uses this class, which relies on preimages *really* not being propagated has its assumptions gutted, doesn't it ?

Reply:
the deafult value for TestConf.preimage_catchup_interval is set to the value:

public Duration preimage_catchup_interval = 100.seconds; 

so none of the existing test fails. We could be set it to much higher.

preimage_catchup_interval is more like a switch in NoPreimageVN, if you want no preimage at all, set it to a high value, but you have to option to customize it and share preimages during nomination.

Reply:
> the default value for TestConf.preimage_catchup_interval is set to the value:
> 
> public Duration preimage_catchup_interval = 100.seconds;
> 
> so none of the existing test fails. We could be set it to much higher.
> 
> preimage_catchup_interval is more like a switch in NoPreimageVN, if you want no preimage at all, set it to a high value(leave it at default), but you have to option to customize it and share ceratain preimages during nomination.

Reply:
Not sure I follow. `NoPreImageVN` should *never* return pre-images if the boolean is set to `false`.
Obviously the rest of the code is unchanged, so it tries to catch-up at a regular interval, but since all validators in the network are `NoPreImageVN`, I would not expect any pre-image to be revealed.

Reply:
> so none of the existing test fails.

They don't fail, but they might start to succeed for the wrong reason.

Reply:
> Not sure I follow. `NoPreImageVN` should _never_ return pre-images if the boolean is set to `false`.
> Obviously the rest of the code is unchanged, so it tries to catch-up at a regular interval, but since all validators in the network are `NoPreImageVN`, I would not expect any pre-image to be revealed.

NoPreImageVN - there are(or will be) at least 3 ways to reveal preimages
1. sharing preimages in advance in a push way
2. sharing preimages during nomination in a pull way
3. preimage catchup

instead of having NoPreImageVN we should allow all those 3 methods to be enabled/disabled (and not just by a bool value), having 

preimage_catchup_interval = 100.seconds;

should be the first step for that

Reply:
> > so none of the existing test fails.
> 
> They don't fail, but they might start to succeed for the wrong reason.

:) I don't get this, if we set 
preimage_catchup_interval = 1000000.seconds;

then the only way to succeed for the wrong reason is that the test actually ran 1000000 seconds and succeeded because of nomination preimage sharing

Reply:
if you have any suggestion how you would like to see this please share in detail and I will implement it that way